In [ ]:
import numpy as np
import pandas as pd

In [2]:
data_raw = pd.read_csv("ppid2_Valence_features.csv")
data_raw.shape

/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_4513/1171576017.py:1: DtypeWarning: Columns (0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80,81,82,83,84,85,86,87,88,89,90,91,92,93,94,95,96,97,98,99,100,101,102,103,104,105,106,107,108,109,110,111,112,113,114,115,116,117,118,119,120,121,122,123,124,125,126,127,128,129,130,131) have mixed types. Specify dtype option on import or set low_memory=False.
  data_raw = pd.read_csv("ppid2_Valence_features.csv")


(5827, 135)

In [3]:
# removing first row
data = data_raw.iloc[1: , :]

# removing where time has null values
data = data[data['Time'].notnull()]

data.head(5)

Time Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC       ACC.1  \
1    0      4  0.15988   0.2456  521.35161  1643.41561  0.06126  6562.72507   
2    1      5    0.161   0.2456  535.30449  1649.63966  0.06126   6553.4292   
3    2      6  0.17014  0.24593   518.3956  1653.90028  0.06126   6552.9059   
4    3      7  0.17014  0.24593  534.68062  1641.58191  0.06742  6556.01749   
5    4      8  0.17014  0.21327   531.5686  1670.96072  0.08616  6577.66903   

       EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55 EEGConn.56  \
1  0.00008  0.00045  ...    0.48763    0.25578    0.01546     0.0168   
2        0        0  ...    0.42995    0.09945    0.04304    0.03403   
3        0        0  ...    0.42875    0.09043    0.05034    0.06566   
4        0        0  ...    0.42099     0.0734    0.05023    0.06085   
5        0  0.00001  ...    0.37806    0.15172    0.40395    0.28963   

  EEGConn.57 EEGConn.58 EEGConn.59 Event Event.1 Event.2  
1    0.13589    0.18419    0.47116   NaN     NaN     NaN  
2    0.05901    0.20058    0.43406   NaN     NaN     NaN  
3     0.0272    0.17715    0.27364   NaN     NaN     NaN  
4    0.02355    0.18303    0.26874   NaN     NaN     NaN  
5    0.02408    0.02495     0.1074   NaN     NaN     NaN  

[5 rows x 135 columns]

In [4]:
data.shape

(5824, 135)

In [5]:
column_list = data.columns.values.tolist()
column_list

['Time',
 'Time.1',
 'EOG',
 'EOG.1',
 'EOG.2',
 'EOG.3',
 'ACC',
 'ACC.1',
 'EDA',
 'EDA.1',
 'EDA.2',
 'EDA.3',
 'EDA.4',
 'EMG',
 'EMG.1',
 'EMG.2',
 'EEGth',
 'EEGth.1',
 'EEGth.2',
 'EEGth.3',
 'EEGth.4',
 'EEGth.5',
 'EEGth.6',
 'EEGth.7',
 'EEGth.8',
 'EEGth.9',
 'EEGth.10',
 'EEGth.11',
 'EEGth.12',
 'EEGth.13',
 'EEGa',
 'EEGa.1',
 'EEGa.2',
 'EEGa.3',
 'EEGa.4',
 'EEGa.5',
 'EEGa.6',
 'EEGa.7',
 'EEGa.8',
 'EEGa.9',
 'EEGa.10',
 'EEGa.11',
 'EEGa.12',
 'EEGa.13',
 'EEGb',
 'EEGb.1',
 'EEGb.2',
 'EEGb.3',
 'EEGb.4',
 'EEGb.5',
 'EEGb.6',
 'EEGb.7',
 'EEGb.8',
 'EEGb.9',
 'EEGb.10',
 'EEGb.11',
 'EEGb.12',
 'EEGb.13',
 'EEGg',
 'EEGg.1',
 'EEGg.2',
 'EEGg.3',
 'EEGg.4',
 'EEGg.5',
 'EEGg.6',
 'EEGg.7',
 'EEGg.8',
 'EEGg.9',
 'EEGg.10',
 'EEGg.11',
 'EEGg.12',
 'EEGg.13',
 'EEGConn',
 'EEGConn.1',
 'EEGConn.2',
 'EEGConn.3',
 'EEGConn.4',
 'EEGConn.5',
 'EEGConn.6',
 'EEGConn.7',
 'EEGConn.8',
 'EEGConn.9',
 'EEGConn.10',
 'EEGConn.11',
 'EEGConn.12',
 'EEGConn.13',
 'EEGConn.14

In [6]:
data.dtypes

Time          object
Time.1        object
EOG           object
EOG.1         object
EOG.2         object
               ...  
EEGConn.58    object
EEGConn.59    object
Event         object
Event.1       object
Event.2       object
Length: 135, dtype: object

### columns:
1 and 2 time (Time is always in chunks, with features calculated over 4 s windows.), \
3-131 features, \
132-134 triggers. 

Column 1: Time start \
Column 2: Time end \
\
Then the features are in 6 types: \
Column 3-6 EOG (eye movements) \
Column 7-8 ACC (accelerometer, head movement) \
Column 9-13 EDA (skin conductance) \
Column 14-16 EMG (face muscle activity) \
Column 17-72 EEG (eeg, in 4 wavebands, eeg theta, alpha, beta, and gamma) \
Column 73-132 EEG connectivity (in 4 wavebands) \
\
Column 133: Trigger values \
Column 134 : Actual time \
Column 135 : Duration (ignored) \
\
Notes on Column 133: \
Sometimes multiple triggers occur in the same window, in which case they are separated by a colon (:) \
the trigger values are too high by 33024 - i.e. subtract from each trigger 33024 to get the right identity \
Trigger to identity conversion. These triggers identify the START of a task: \
223: Training, negative valence \
224: Training, positive valence \
233: Testing, negative valence \
234: Testing, positive valence \
Trigger 7 identifies the END of a task \

In [7]:
# how many NaN values in every column
data.isnull().sum(axis = 0)

Time             0
Time.1           0
EOG              0
EOG.1            0
EOG.2            0
              ... 
EEGConn.58       0
EEGConn.59       0
Event         5157
Event.1       5157
Event.2       5157
Length: 135, dtype: int64

In [8]:
# Row where trigger values are not null
data[data['Event'].notnull()]

Time  Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
106      105     109  0.08397  0.17754  532.01239  1651.24618  0.13116   
198      197     201  0.06053  0.18658  535.46658  1658.04357  0.11674   
199      198     202  0.07965  0.19993  527.66078  1653.63064  0.05039   
210      209     213  0.02848  0.09778  527.55139  1649.96149   0.0356   
217      216     220  0.06589  0.14229  529.07506  1649.17702  0.06266   
...      ...     ...      ...      ...        ...         ...      ...   
5765  5764.0  5768.0  0.04436  0.20887  526.67788  1644.21708  0.03057   
5781  5780.0  5784.0  0.03627  0.11686  527.96698  1648.71264  0.11557   
5783  5782.0  5786.0  0.03627  0.11651  524.25135  1653.06177  0.16028   
5784  5783.0  5787.0  0.03493   0.1533  528.04897  1661.89567  0.16028   
5811  5810.0  5814.0  0.03251  0.15251  529.80917  1652.22609  0.02847   

           ACC.1  EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55  \
106   6466.21863    0  0.00001  ...    0.24501    0.07439    0.15514   
198    6635.4539    0  0.00001  ...     0.2749    0.39998     0.0503   
199   6611.59804    0  0.00001  ...    0.81818    0.13074     0.0557   
210   6520.06752    0        0  ...    0.35827    0.20248    0.13175   
217    6525.1543    0  0.00004  ...    0.04188    0.28173    0.42206   
...          ...  ...      ...  ...        ...        ...        ...   
5765  6490.01525  0.0  0.00001  ...    0.02219    0.16778    0.25196   
5781   6401.3176  0.0      0.0  ...    0.11492    0.24175    0.05891   
5783  6428.85437  0.0      0.0  ...     0.0464    0.16571    0.38692   
5784  6456.51772  0.0      0.0  ...    0.09351     0.1209    0.56577   
5811  6503.73269  0.0      0.0  ...    0.20457    0.50569    0.36033   

     EEGConn.56 EEGConn.57 EEGConn.58 EEGConn.59              Event  \
106     0.06534    0.12709    0.07207    0.02448              33029   
198     0.10911     0.2435    0.20171    0.19819              33030   
199     0.06662    0.12553    0.23329    0.04058              33124   
210     0.09717    0.05942    0.28975    0.03434  33031:33033:33029   
217     0.27317    0.14907    0.05713    0.06576              33030   
...         ...        ...        ...        ...                ...   
5765    0.15766     0.3969    0.25981    0.41914              33032   
5781    0.07146    0.17676    0.18293    0.51784        33033:33029   
5783    0.14389    0.06218    0.24518    0.44891              33030   
5784    0.36061    0.17068    0.20694    0.42465  33135:33256:33257   
5811    0.08966     0.3681    0.29534    0.66486        33031:33032   

                               Event.1                  Event.2  
106                             108.73                        0  
198                            200.156                        0  
199                            201.158                        0  
210      212.16400:212.66400:212.69800  0.00000:0.00000:0.00000  
217                             219.77                        0  
...                                ...                      ...  
5765                          5767.332                        0  
5781             5783.49200:5783.52600          0.00000:0.00000  
5783                           5785.26                        0  
5784  5786.26000:5786.26200:5786.26400  0.00000:0.00000:0.00000  
5811             5813.26200:5813.77800          0.00000:0.00000  

[667 rows x 135 columns]

In [9]:
# Training - Negative
data[data['Event'].str.contains('33247', na=False)]

Time Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
291    290    294  0.04307  0.15639  527.64843  1648.30522  0.03447   
330    329    333  0.14531  0.17405  519.03303   1660.4022  0.07649   
360    359    363  0.04086  0.19042  529.78824  1650.28172  0.21163   
399    398    402  0.06836  0.16566   533.0209  1651.15145  0.24142   
431    430    434  0.13613  0.20408  524.28051  1660.82444  0.03159   
446    445    449  0.04194  0.14465  526.39621  1649.35849  0.04288   
515    514    518  0.04914  0.15438  527.52378  1649.91815  0.03402   
557    556    560  0.05509  0.15631  529.64183  1654.70007  0.04934   
631    630    634  0.03721  0.17628    527.916   1647.4125  0.04584   
678    677    681   0.0419  0.15813  529.94111  1650.06347  0.03802   
728    727    731  0.08017  0.14668    528.652  1654.54942  0.18274   
759    758    762  0.04203  0.15738  528.25491  1640.47899  0.04295   
828    827    831  0.05733  0.17785  525.91743  1655.40933  0.11661   
843    842    846  0.03379  0.16047   529.0382  1653.24649    0.043   
898    897    901  0.04473  0.14252  533.20244  1652.23643  0.04642   
915    914    918   0.0674  0.15818  525.20282  1647.61129  0.14495   
974    973    977  0.03958  0.15735  529.91507  1658.02043   0.0478   
1018  1017   1021  0.05887  0.15407  532.09396   1657.1844  0.04574   
1088  1087   1091  0.13078  0.24648  527.96083  1655.29899  0.04427   
1104  1103   1107   0.1219  0.23309  522.11391  1665.09528  0.05429   
1119  1118   1122  0.06135  0.18878  529.16896  1639.19303  0.16721   
1164  1163   1167  0.16293  0.22937  530.58636  1649.97174   0.0496   
1230  1229   1233  0.11047  0.20071  527.22839  1663.52725  0.07134   
1251  1250   1254  0.08558  0.20882  526.57061  1653.68739  0.19739   

           ACC.1       EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55  \
291   6520.46483         0  0.00002  ...    0.21702    0.14526    0.25247   
330   6550.98072         0        0  ...    0.24145    0.12684    0.07766   
360   6480.76176         0        0  ...    0.28579    0.38258    0.43549   
399   6505.51685         0  0.00007  ...    0.10523    0.16946    0.42892   
431   6461.32961         0  0.00004  ...    0.02516    0.13682     0.3201   
446   6495.47203         0  0.00007  ...    0.11822     0.1458    0.24954   
515   6481.05937         0  0.00011  ...    0.08989    0.05591    0.32822   
557   6500.02922         0  0.00001  ...    0.04572    0.14242    0.28657   
631   6497.72887         0        0  ...    0.07014    0.14001    0.25183   
678   6479.52223         0  0.00001  ...    0.17207     0.0184    0.29019   
728   6479.80952         0  0.00002  ...    0.03891    0.10774    0.76956   
759   6487.67885         0        0  ...    0.16243    0.03083    0.50822   
828   6486.59865  -0.00001  0.00002  ...    0.08399    0.33804    0.17856   
843   6487.48169         0  0.00001  ...    0.17101    0.13633    0.29912   
898   6520.64785         0  0.00003  ...    0.01391    0.01315    0.26941   
915   6566.45604         0  0.00006  ...    0.59008    0.41131    0.39596   
974   6476.56206         0  0.00006  ...    0.12374    0.14103    0.47078   
1018  6544.85577         0  0.00002  ...     0.4126    0.38457    0.07005   
1088  6469.11183         0  0.00004  ...    0.06007    0.39411    0.24626   
1104  6515.94785         0  0.00013  ...    0.07795    0.05095    0.16715   
1119  6477.82374         0  0.00055  ...    0.01848    0.07219    0.13416   
1164  6503.14972         0        0  ...    0.00201    0.09394    0.28242   
1230  6508.73544         0        0  ...     0.4622    0.00298    0.26204   
1251  6463.20736         0  0.00048  ...     0.2164     0.1405     0.4618   

     EEGConn.56 EEGConn.57 EEGConn.58 EEGConn.59              Event  \
291      0.0632    0.14337    0.00115    0.32136  33125:33246:33247   
330     0.01367    0.14092     0.0998    0.03128  33127:33245:33247   
360     0.22268    0.01473     0.0199    0.12771  33125:33246:33247   
399     0.17696    0

In [10]:
# Training - Positive

data[data['Event'].str.contains('33248', na=False)]

Time Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
276    275    279  0.08529  0.19562   525.9664  1651.58225  0.05048   
311    310    314  0.05151  0.17885   527.8263  1643.30627  0.14037   
377    376    380  0.13172  0.21739  531.81364  1671.15264  0.23671   
415    414    418  0.04161  0.17612  529.33468  1646.43311  0.12148   
475    474    478  0.03499  0.13956  529.37823  1656.28647  0.03973   
499    498    502  0.05392  0.14528  530.56574  1662.42407  0.08494   
533    532    536  0.18046  0.25314  528.52575  1655.06686  0.03391   
603    602    606   0.0448  0.17729  525.32587   1653.1928  0.20664   
647    646    650  0.04539  0.19289  526.87817  1652.32046   0.0471   
663    662    666  0.05764  0.18325  528.71251  1643.72851  0.03827   
744    743    747  0.10272  0.20894  528.81577  1650.23656  0.04081   
777    776    780  0.06315  0.15369  524.09065  1640.43457  0.19204   
791    790    794  0.02982  0.19106  527.90979  1649.64041   0.0414   
860    859    863  0.03556  0.22091  523.55287  1651.75915  0.04071   
930    929    933  0.03422  0.17758  526.76335  1641.86825   0.2451   
945    944    948  0.03279  0.16796  524.90402  1647.32644  0.03914   
959    958    962  0.04028  0.16215  528.89802  1655.97136  0.07373   
994    993    997  0.04857  0.15925  528.36975  1654.50956  0.18657   
1060  1059   1063  0.03574  0.15862  529.92753  1658.05541   0.1417   
1074  1073   1077  0.02921  0.18165  530.08456  1646.27315  0.05118   
1135  1134   1138  0.03275  0.20455  525.99935  1646.74139  0.05549   
1184  1183   1187  0.06796  0.17104  523.16165  1653.13331  0.18353   
1216  1215   1219  0.05426  0.20382  528.77274  1656.48633  0.26955   
1265  1264   1268  0.04605  0.15856   527.5871  1651.47215  0.04061   

           ACC.1       EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55  \
276   6459.28836         0  0.00001  ...    0.01789    0.05837    0.29912   
311   6523.02835         0  0.00031  ...    0.06379    0.07516    0.27044   
377   6519.82102  -0.00001  0.00006  ...    0.20517    0.05392    0.44147   
415   6500.04767         0  0.00006  ...    0.16599    0.05031    0.32471   
475   6514.90613         0        0  ...    0.10168    0.12951    0.27054   
499   6511.83045         0  0.00046  ...    0.40275    0.00359    0.54102   
533   6494.08954         0        0  ...    0.10283    0.11934    0.40284   
603   6504.64318         0  0.00005  ...    0.00962    0.05227     0.5689   
647   6514.57355         0  0.00002  ...      0.209    0.05706     0.2005   
663   6501.29403         0  0.00003  ...    0.17765    0.05451    0.08036   
744   6483.38135         0        0  ...    0.01482    0.33293    0.15917   
777   6500.74497         0  0.00003  ...    0.11415    0.12483    0.39172   
791   6502.57401         0  0.00001  ...    0.15117    0.31913    0.46067   
860   6472.98922         0  0.00073  ...    0.27951    0.16447    0.35029   
930   6527.17155         0        0  ...    0.04843    0.00229     0.5012   
945   6457.99805         0  0.00005  ...    0.00279    0.19592     0.4249   
959   6517.23297         0        0  ...    0.03282    0.08583    0.37111   
994   6602.17124  -0.00001        0  ...    0.18164    0.21395    0.37814   
1060  6517.72424         0  0.00002  ...    0.01697     0.2057    0.38516   
1074  6450.67539         0  0.00003  ...    0.03008    0.14706    0.60798   
1135   6495.1529         0  0.00005  ...    0.07581    0.03468    0.43671   
1184  6524.31709         0  0.00024  ...    0.26289    0.04318    0.19098   
1216  6508.40367         0        0  ...    0.11179    0.32802    0.30102   
1265  6467.68879         0        0  ...    0.01945    0.25841    0.30293   

     EEGConn.56 EEGConn.57 EEGConn.58 EEGConn.59              Event  \
276     0.19441    0.27426    0.60677    0.14221  33128:33245:33248   
311       0.097    0.00896    0.00839    0.36745  33126:33246:33248   
377      0.1779    0.09689     0.0082    0.48181  33126:33246:33248   
415     0.05102    0

In [11]:
# Testing - Negative

data[data['Event'].str.contains('33257', na=False)]

Time  Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
1491    1490    1494  0.04975  0.14598  524.55688  1653.47183  0.05797   
1547    1546    1550  0.04173  0.18545  526.75122  1652.10352  0.12971   
1659    1658    1662  0.05063  0.21209  528.39099  1641.40343   0.1223   
1742    1741    1745   0.0386  0.18553  529.68393   1655.3671  0.14063   
1961    1960    1964  0.10332   0.2381  526.32548  1650.02833  0.10101   
2029    2028    2032  0.03192  0.17481  527.59151  1651.14483  0.15999   
2140    2139    2143  0.04723  0.20369  528.64549  1653.00139  0.27857   
2277    2276    2280  0.10675  0.24025  526.51209  1640.30193   0.2939   
2312    2311    2315  0.15001  0.25631  529.94923  1660.38128   0.2031   
2403    2402    2406  0.04872  0.21078  529.08346  1665.97577  0.22072   
2514    2513    2517  0.04212    0.144  526.50118  1648.71132  0.11294   
2627    2626    2630  0.02796   0.1779   528.7757  1653.29413  0.08592   
2698    2697    2701  0.03743  0.15354  529.06812  1650.67535  0.15668   
2896    2895    2899  0.04822  0.16185  526.12794  1646.49238  0.36281   
3023    3022    3026  0.03341  0.16164  528.55225  1649.86688  0.06468   
3111    3110    3114  0.03964  0.15878  527.91146  1658.37331   0.2375   
3279    3278    3282  0.04084  0.16049  528.58253  1652.69929  0.27173   
3343    3342    3346  0.10816  0.21355  528.21936  1649.89163   0.0941   
3466    3465    3469  0.03931  0.16669  527.26663  1658.47586  0.25214   
3520    3519    3523  0.08521  0.19741  527.86439  1646.59649  0.13472   
3607    3606    3610  0.25538   0.2251  529.12543   1650.6884  0.23995   
3656    3655    3659  0.03695  0.06014  528.84204  1652.13748  0.04774   
3954    3953    3957  0.05443  0.21614  527.90271  1652.21402  0.11024   
4049    4048    4052  0.06921  0.22323  529.80192  1657.89789  0.21367   
4078    4077    4081  0.04469   0.1954  527.37925  1652.05738  0.16751   
4110  4109.0  4113.0  0.04345  0.18892   527.9415  1652.72874  0.24705   
4175  4174.0  4178.0  0.05127  0.22916  528.77914  1648.83543  0.27572   
4247  4246.0  4250.0  0.04928   0.2205  526.95902  1653.12171  0.16101   
4287  4286.0  4290.0  0.03862  0.13713  528.27301  1646.47512  0.17456   
4402  4401.0  4405.0  0.04716  0.21249  524.99801  1642.38162  0.23783   
4497  4496.0  4500.0  0.07327  0.26048  528.09637  1649.14641  0.24132   
4587  4586.0  4590.0  0.03923  0.24746  530.27782  1649.55109  0.08969   
4615  4614.0  4618.0  0.03413  0.16435  528.87141  1651.22177  0.13216   
4646  4645.0  4649.0  0.04328  0.22223   528.6351  1650.38975  0.25916   
4719  4718.0  4722.0  0.05674  0.24417  532.22522  1644.28296   0.4094   
4928  4927.0  4931.0   0.0382  0.23575  529.94196  1657.43888  0.47424   
4957  4956.0  4960.0  0.02764  0.21861  527.91197  1650.02138  0.24139   
5036  5035.0  5039.0  0.05558  0.22987  527.82715  1652.86659  0.20027   
5077  5076.0  5080.0  0.05643  0.26216  527.04088  1649.96831  0.37074   
5109  5108.0  5112.0  0.04652  0.24405  527.21968  1648.97295  0.38234   
5150  5149.0  5153.0   0.0322  0.16522  528.10707  1654.58914  0.17945   
5357  5356.0  5360.0  0.05983  0.22997  527.88873  1648.79688  0.16358   
5400  5399.0  5403.0  0.04508  0.19688  529.98781  1653.66694  0.17323   
5430  5429.0  5433.0  0.06436  0.18409  526.09741  1648.86243   0.1164   
5501  5500.0  5504.0   0.0756  0.25204  526.30573    1650.394  0.20129   
5549  5548.0  5552.0  0.07972  0.22842  528.95696  1654.63675  0.23191   
5587  5586.0  5590.0  0.06213  0.22594  528.82195  1655.46243  0.36938   
5784  5783.0  5787.0  0.03493   0.1533  528.04897  1661.89567  0.16028   

           ACC.1       EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55  \
1491   6502.0023  -0.00002   0.0012  ...    0.70751     0.0112    0.17988   
1547  6618.69222         0  0.00008  ...    0.06074    0.06915    0.19776   
1659  6517.78439         0        0  ...      0.448    0.03916    0.16267   
1742  6591.12141         0  0.00007  ...          0    0.24382    0.11507

In [12]:
# Testing - Positive

data[data['Event'].str.contains('33258', na=False)]

Time  Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
1386    1385    1389  0.06317  0.21772  526.39146  1646.68596  0.03357   
1613    1612    1616  0.04165  0.20314  530.18703  1659.96148  0.20347   
1705    1704    1708  0.07835  0.22699  526.16651  1660.13599  0.23369   
1792    1791    1795  0.03636  0.13301  527.81482  1650.97286  0.19155   
1828    1827    1831  0.03412  0.26456  527.78406  1654.42059  0.15431   
1869    1868    1872  0.03991  0.16447  528.63029  1657.18354  0.22919   
1913    1912    1916  0.03784  0.21798  529.02754  1657.62359  0.22065   
2094    2093    2097  0.04035  0.20809   531.7399  1673.97747  0.22369   
2182    2181    2185  0.07313  0.24699  529.64643  1655.52294  0.23266   
2238    2237    2241  0.05325  0.18102  528.46019   1654.6817  0.33581   
2365    2364    2368   0.0406  0.17363  527.65071  1654.51089  0.29379   
2459    2458    2462  0.04618  0.17644  529.60926  1662.12826  0.17805   
2569    2568    2572  0.04981  0.16047  526.65468  1652.67987  0.22028   
2749    2748    2752  0.06696  0.20406  522.08333  1639.11299  0.06135   
2829    2828    2832  0.20106  0.23259  522.89053  1658.65036  0.25858   
2959    2958    2962   0.0458  0.18414  524.80895  1656.31327  0.30244   
3069    3068    3072  0.06166  0.19141   525.6283  1647.27731  0.32293   
3148    3147    3151  0.04755  0.22931  528.26995  1655.51026  0.36175   
3206    3205    3209  0.05022  0.16373  528.37729  1651.72301  0.14687   
3431    3430    3434  0.03729  0.15774  528.23424  1647.31291  0.05381   
3564    3563    3567  0.04851  0.18495  530.37676  1650.54556  0.19026   
3749    3748    3752  0.12783   0.2418  524.94792  1656.82717  0.25893   
3787    3786    3790  0.11176  0.18124  530.12154  1651.78174  0.08561   
3816    3815    3819  0.04111  0.23538  527.62584  1650.16755     0.11   
3847    3846    3850  0.05949  0.19816  524.17629  1649.39127  0.29896   
3900    3899    3903  0.05363  0.20593  529.49057  1649.83661  0.26197   
4014    4013    4017  0.10835  0.13809  524.17812  1649.25403  0.21811   
4144  4143.0  4147.0   0.0734  0.24503  529.38165  1647.14617  0.24005   
4212  4211.0  4215.0  0.07206  0.24443  526.16293  1651.71405  0.19994   
4345  4344.0  4348.0  0.05958   0.2572   517.7524  1653.39954  0.15296   
4448  4447.0  4451.0  0.14431  0.26876  527.07487  1646.94619  0.22921   
4533  4532.0  4536.0  0.04396  0.26572  525.95172  1644.38723  0.32374   
4679  4678.0  4682.0   0.0374  0.17584  528.27949  1654.21867  0.24271   
4765  4764.0  4768.0  0.04628   0.2233  528.79386  1650.33429  0.25406   
4816  4815.0  4819.0  0.08668  0.24564  525.16994  1656.60297  0.24258   
4850  4849.0  4853.0   0.0503  0.23572  528.23456  1652.74903  0.34138   
4894  4893.0  4897.0  0.12307  0.17714  528.24639  1655.63273   0.1268   
4987  4986.0  4990.0  0.05419  0.15801  526.79356  1644.94201  0.18881   
5174  5173.0  5177.0  0.05797  0.13966  525.39067  1644.75916  0.17324   
5200  5199.0  5203.0  0.06031  0.30337  529.33589  1650.27203   0.3643   
5247  5246.0  5250.0  0.06202  0.15523  527.79133  1645.50599   0.1623   
5273  5272.0  5276.0  0.04526  0.16868  530.46676  1651.71799  0.26362   
5309  5308.0  5312.0  0.03439  0.17443  528.31881  1651.06262  0.05652   
5470  5469.0  5473.0  0.04196  0.14893  531.28897  1653.25368   0.1789   
5620  5619.0  5623.0    0.062  0.22701   528.2784  1647.69457  0.41155   
5667  5666.0  5670.0     0.06  0.19394  527.51239  1650.79699  0.22106   
5714  5713.0  5717.0  0.04741  0.25263  526.69526  1645.43386  0.21042   
5749  5748.0  5752.0  0.09198  0.20428  526.19168  1645.57965  0.15369   

           ACC.1       EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55  \
1386  6511.26579         0  0.00003  ...    0.72877    0.04928    0.39149   
1613  6601.77679  -0.00002  0.00126  ...    0.04597    0.04857    0.30465   
1705  6514.19806         0  0.00007  ...    0.04662    0.14247    0.37657   
1792  6627.81224         0  0.00001  ...    0.01419    0.27447    0.34026

In [13]:
# Trigger ends

data[data['Event'].str.contains('33031', na=False)]

Time  Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
210      209     213  0.02848  0.09778  527.55139  1649.96149   0.0356   
229      228     232  0.02767  0.13325  527.59964  1657.03033  0.02761   
247      246     250  0.02225  0.12874  528.98606  1646.20776  0.02284   
270      269     273  0.03387   0.1476  529.83433  1650.16155  0.03354   
287      286     290  0.03336  0.14551  527.85654  1658.94438  0.03124   
...      ...     ...      ...      ...        ...         ...      ...   
5647  5646.0  5650.0  0.02901  0.15213  528.34839  1652.47307  0.07342   
5694  5693.0  5697.0  0.02411  0.14216  527.98252   1654.4142  0.06862   
5729  5728.0  5732.0   0.0147  0.05429  527.77479  1650.84935  0.03241   
5764  5763.0  5767.0  0.01913  0.06253  529.62598  1648.91512  0.03166   
5811  5810.0  5814.0  0.03251  0.15251  529.80917  1652.22609  0.02847   

           ACC.1  EDA    EDA.1  ... EEGConn.53 EEGConn.54 EEGConn.55  \
210   6520.06752    0        0  ...    0.35827    0.20248    0.13175   
229   6467.97309    0  0.00005  ...    0.06028    0.14772    0.19554   
247   6493.27511    0  0.00004  ...     0.0105    0.08807    0.60728   
270   6437.42658    0        0  ...    0.17172    0.00097     0.4092   
287   6518.29474    0        0  ...    0.01879     0.1343    0.31981   
...          ...  ...      ...  ...        ...        ...        ...   
5647  6494.44843  0.0  0.00002  ...    0.00259    0.15465    0.41478   
5694  6498.09062  0.0      0.0  ...    0.10734    0.13069    0.42102   
5729  6496.22257  0.0  0.00023  ...    0.00804    0.19745    0.49048   
5764  6488.36373  0.0  0.00001  ...    0.06741    0.19848    0.32124   
5811  6503.73269  0.0      0.0  ...    0.20457    0.50569    0.36033   

     EEGConn.56 EEGConn.57 EEGConn.58 EEGConn.59              Event  \
210     0.09717    0.05942    0.28975    0.03434  33031:33033:33029   
229      0.1471    0.04204    0.00402    0.29212              33031   
247     0.48295    0.11828    0.01579      0.515              33031   
270     0.32275     0.0559    0.03842    0.51788              33031   
287     0.18148    0.21457    0.07417    0.20328              33031   
...         ...        ...        ...        ...                ...   
5647    0.23192    0.22881    0.16313    0.47673        33031:33032   
5694    0.17151    0.06628     0.0499      0.501              33031   
5729    0.38343    0.31892    0.19585    0.08032              33031   
5764    0.10682    0.41776    0.24157    0.61151              33031   
5811    0.08966     0.3681    0.29534    0.66486        33031:33032   

                            Event.1                  Event.2  
210   212.16400:212.66400:212.69800  0.00000:0.00000:0.00000  
229                         231.778                        0  
247                         249.622                        0  
270                         272.954                        0  
287                          289.95                        0  
...                             ...                      ...  
5647          5649.45400:5649.95400          0.00000:0.00000  
5694                       5696.918                        0  
5729                       5731.808                        0  
5764                       5766.814                        0  
5811          5813.26200:5813.77800          0.00000:0.00000  

[159 rows x 135 columns]

In [14]:
data['Type'] = 'random'

In [15]:
data['Valence'] = 'random'

In [16]:
# how many NaN values in every column
data.isnull().sum(axis = 0)

Time          0
Time.1        0
EOG           0
EOG.1         0
EOG.2         0
           ... 
Event      5157
Event.1    5157
Event.2    5157
Type          0
Valence       0
Length: 137, dtype: int64

In [17]:
data.head()

Time Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC       ACC.1  \
1    0      4  0.15988   0.2456  521.35161  1643.41561  0.06126  6562.72507   
2    1      5    0.161   0.2456  535.30449  1649.63966  0.06126   6553.4292   
3    2      6  0.17014  0.24593   518.3956  1653.90028  0.06126   6552.9059   
4    3      7  0.17014  0.24593  534.68062  1641.58191  0.06742  6556.01749   
5    4      8  0.17014  0.21327   531.5686  1670.96072  0.08616  6577.66903   

       EDA    EDA.1  ... EEGConn.55 EEGConn.56 EEGConn.57 EEGConn.58  \
1  0.00008  0.00045  ...    0.01546     0.0168    0.13589    0.18419   
2        0        0  ...    0.04304    0.03403    0.05901    0.20058   
3        0        0  ...    0.05034    0.06566     0.0272    0.17715   
4        0        0  ...    0.05023    0.06085    0.02355    0.18303   
5        0  0.00001  ...    0.40395    0.28963    0.02408    0.02495   

  EEGConn.59 Event Event.1 Event.2    Type Valence  
1    0.47116   NaN     NaN     NaN  random  random  
2    0.43406   NaN     NaN     NaN  random  random  
3    0.27364   NaN     NaN     NaN  random  random  
4    0.26874   NaN     NaN     NaN  random  random  
5     0.1074   NaN     NaN     NaN  random  random  

[5 rows x 137 columns]

In [18]:
# Labeling the data
train_neg = '33247'
train_pos = '33248'
test_neg = '33257'
test_pos = '33258'
end = '33031'

# non null data
non_null_data = data[data['Event'].notnull()]

In [19]:
# cast into string
non_null_data['Event'] = non_null_data['Event'].astype(str)
non_null_data['Event.1'] = non_null_data['Event.1'].astype(str)

/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_67192/3744656094.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_null_data['Event'] = non_null_data['Event'].astype(str)
/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_67192/3744656094.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  non_null_data['Event.1'] = non_null_data['Event.1'].astype(str)


In [20]:
# iterating all to get start time and end time
train_neg_list = []
train_pos_list = []
test_neg_list = []
test_pos_list = []

tmp_start_time = 0
tmp_end_time = 0
# 0 = not occuring, 1 = train_neg, 2 = train_pos, 3 = test_neg, 4 = test_pos
types = 0

for index, row in non_null_data.iterrows():
    event = row['Event'].split(':')
    time = row['Event.1'].split(':')
    
    if types == 0:
        idx = -1
        # find the start time
        if train_neg in event:
            idx = event.index(train_neg)
            types = 1
        if train_pos in event:
            idx = event.index(train_pos)
            types = 2
        if test_neg in event:
            idx = event.index(test_neg)
            types = 3
        if test_pos in event:
            idx = event.index(test_pos)
            types = 4
        
        if idx != -1:
            tmp_start_time = time[idx]
    else:
        # find ending
        if end in event:
            idx = event.index(end)
            tmp_end_time = time[idx]
            
            duration = (tmp_start_time, tmp_end_time)
            if types == 1:
                train_neg_list.append(duration)
            elif types == 2:
                train_pos_list.append(duration)
            elif types == 3:
                test_neg_list.append(duration)
            else:
                test_pos_list.append(duration)
            
            # reset everything
            tmp_start_time = 0
            tmp_start_time = 0
            types = 0

In [21]:
print(len(train_neg_list))
print(len(train_pos_list))
print(len(test_neg_list))
print(len(test_pos_list))

24
24
48
48


In [22]:
# cast into float
data['Time'] = data['Time'].astype(float)
data['Time.1'] = data['Time.1'].astype(float)

In [23]:
# labeling based on (start, end) tuples

# labeling train negatives
for start,end in train_neg_list:
    start = float(start)
    end = float(end)
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Training'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Negative'
    
# labeling train positives
for start,end in train_pos_list:
    start = float(start)
    end = float(end)
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Training'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Positive'
    
# labeling test negatives
for start,end in test_neg_list:
    start = float(start)
    end = float(end)
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Testing'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Negative'
    
# labeling test positives
for start,end in test_pos_list:
    start = float(start)
    end = float(end)
    
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Type'] = 'Testing'
    data.loc[((data['Time'] > start) & (data['Time.1'] < end)), 'Valence'] = 'Positive'

In [24]:
# clean data. only the one that have labels

clean_data = data[data['Type'] != 'random']

In [25]:
pd.set_option('display.max_columns', None)

clean_data.head(5)

Time  Time.1      EOG    EOG.1      EOG.2       EOG.3      ACC  \
280  279.0   283.0  0.03501  0.18852   528.3009  1653.59465  0.08911   
281  280.0   284.0  0.03269  0.18852  530.00618  1663.87925  0.07023   
282  281.0   285.0  0.03522  0.16972  527.78995  1650.26793  0.03413   
283  282.0   286.0  0.03522  0.16972  527.84376  1645.88954  0.02725   
284  283.0   287.0  0.03608  0.16972  530.29163  1653.78549  0.02725   

          ACC.1 EDA    EDA.1    EDA.2     EDA.3    EDA.4      EMG    EMG.1  \
280  6520.83036   0   0.0003  0.01733  -0.00001  0.00001   3.9851  4.16329   
281  6531.95043   0  0.00003  0.00579         0  0.00001  3.69995  3.90775   
282  6524.41338   0  0.00001  0.00305         0  0.00001   3.1142  3.74782   
283  6522.16183   0  0.00004  0.00617   0.00001  0.00001  3.11963  3.70944   
284  6520.32162   0  0.00003  0.00531         0  0.00001  3.33145  3.79952   

       EMG.2    EEGth  EEGth.1  EEGth.2  EEGth.3  EEGth.4  EEGth.5  EEGth.6  \
280  3.40618  7.04455  4.86199  3.21675  5.75429  6.44093  5.61089   3.2169   
281  3.15758  7.82496  3.74952  3.76349  6.06088  6.00393  5.40851  1.93406   
282  3.16962  7.98087  4.33689  4.06249  6.83941  6.33279  5.59029  1.87157   
283  3.17337  8.28311  4.41303  3.94858  6.66202  6.25864  5.84345   1.9472   
284  3.22674  8.63026  4.19899  3.83346   6.2585  6.13271  5.64778  2.44353   

     EEGth.7  EEGth.8  EEGth.9 EEGth.10 EEGth.11 EEGth.12 EEGth.13      EEGa  \
280   2.5645    4.762  4.03971  3.42983  2.04598  4.81479  7.35615   5.77247   
281  2.04336  4.48601  4.15952  2.93264   1.6905  4.70753  7.22885   5.80802   
282   1.8956  4.65682  4.86761  3.44949  1.87841   4.6306  7.57083   9.15003   
283  2.04153   4.5926  4.54632  3.57249  2.08678  4.92243  7.74617    9.4604   
284  2.67223  4.29046  4.32416  3.52351  2.05628  4.53448  7.48535  10.19491   

      EEGa.1   EEGa.2   EEGa.3    EEGa.4   EEGa.5   EEGa.6   EEGa.7   EEGa.8  \
280  5.57593  4.59295  7.39446   7.39746  8.47643  4.48683  2.75862  5.20173   
281  4.26961  4.18997  6.70635   6.64888  6.66284  3.21255   2.9131  4.26131   
282  5.73762  5.47857  8.80649   8.76481  8.63043   2.8717  3.18623  4.57234   
283  5.51253   5.3575  8.43309   8.44539  8.27111  2.89486  3.30245   4.8055   
284   6.2503  5.47303  8.89868  10.27402   9.2888  3.38885  3.09618    4.918   

      EEGa.9  EEGa.10  EEGa.11  EEGa.12   EEGa.13     EEGb   EEGb.1   EEGb.2  \
280   6.3108  4.46472  2.72489   4.2044  10.16926  5.21583  7.17967  2.69527   
281  5.23405  3.52305  2.29329  3.81119   8.38724  5.37571  3.27578  2.58342   
282  7.01924  4.70875  2.58103  3.88574  11.85267   5.6117  3.09253  2.72783   
283   6.7022  5.11825  3.04849  3.50387  11.36607  5.34157  3.05257  2.58004   
284  8.05664  5.53824  3.24834  3.42464  13.33634  5.37738  2.86439  2.58095   

      EEGb.3   EEGb.4   EEGb.5   EEGb.6   EEGb.7   EEGb.8   EEGb.9  EEGb.10  \
280  5.98089   9.7621   4.1087  5.83284  2.41022  3.76727  3.08318  2.31651   
281  3.88451  6.19128  4.11012  2.51016  2.33732   2.4323  2.98238  2.11472   
282  4.16805  5.54339  4.08376  2.29635   2.1814  2.77989  3.22493  2.17092   
283  4.07061  5.60008  3.90344  2.33227  2.20498   2.8184  3.18053  2.10575   
284  4.02496  5.43062  3.97381  2.26897  2.14328  2.82263  2.93255  2.20143   

     EEGb.11  EEGb.12  EEGb.13     EEGg   EEGg.1   EEGg.2   EEGg.3   EEGg.4  \
280  1.52876  4.91268  6.57246  3.12899  6.12816  1.47342  4.26038  8.11057   
281   1.4377  3.22846  6.83003  2.50325  2.42307  1.17426  2.22927  4.88182   
282  1.39118  3.39798  7.12267  2.46459  2.01071  1.13092  2.02038  4.42747   
283  1.30571  3.40154  7.31966  2.54512  1.88962  1.11229  1.90157  4.32785   
284  1.37513  3.31449  6.99063   2.6607  1.78469  1.11821  1.83363  4.08216   

      EEGg.5   EEGg.6   EEGg.7   EEGg.8   EEGg.9  EEGg.10  EEGg.11  EEGg.12  \
280  2.51996  5.63534  1.27287  2.72814  1.67304  1.53843  0.71795  4.34951   
281  1.84993  1.70475  1.00018   1.1424  1.31856   0.8742  0.49342  2.38991   

In [26]:
# casting feature values to float, type and valence to string, removing events

feature_cols = list(clean_data.iloc[:,0:132])# Creates list of all column headers
clean_data[feature_cols] = clean_data[feature_cols].astype(float)

target_cols = ['Type','Valence']
clean_data[target_cols] = clean_data[target_cols].astype(str)

clean_data.drop(columns=['Event', 'Event.1', 'Event.2'], inplace=True)

/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_67192/1750340855.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_data[feature_cols] = clean_data[feature_cols].astype(float)
/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_67192/1750340855.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  clean_data[target_cols] = clean_data[target_cols].astype(str)
/var/folders/f4/6_9y1nb93x72_74wsk9vg8qh0000gn/T/ipykernel_67192/1750340855.py:9: SettingWithCopyWarning: 
A value is trying to 

In [27]:
clean_data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1584 entries, 280 to 5810
Columns: 134 entries, Time to Valence
dtypes: float64(132), object(2)
memory usage: 1.6+ MB


In [28]:
# save clean data to a csv
clean_data.to_csv('clean_data.csv', encoding='utf-8', index=False)

In [7]:
clean_data = pd.read_csv("clean_data.csv")
clean_data.shape

(1584, 134)